### DISCUSSION TITLE
Error : The memory block 'Elsa.Expressions.Models.MemoryBlockReference' does not exist

### DISCUSSION THREAD

**Discussion by mremadevi** [elsa-workflows/elsa-core#5290] [Q&A]
The memory block 'Elsa.Expressions.Models.MemoryBlockReference' does not exist for below workflow (starting new discussion with more details)

**Context**

1.  Custom activity 'ActivityWithOutput1' is defined in .Net 8
2. Workflow is designed from designer and stored as Json
3. C# code loads JSON workflow and  runs workflow using below code

 
 var result = await context.ServiceProvider.GetRequiredService<IWorkflowRunner>().RunAsync(workflow, new RunWorkflowOptions
  {           
      CancellationTokens = cancellationToken ,
      
  }, cancellationToken).ConfigureAwait(false);

4.  Workflow is not expecting any input . But uses variable to store result of  'ActivityWithOutput1'   for usage in 'If'. Tried for  accessing result with Output.LastResult as well as Variables.Get(). Both gives same error . 

5.  Variable count shows as zero when loading workflow json even though my workflow json has variables defined
6.  'ActivityWithOutput1'  does set output correctly  and here is what I see from my postgre persistence in AcitivityExecutionRecords

ActivityWithOutput1	2024-04-26 14:38:30.058 -0500	false	Completed	2024-04-26 14:38:34.985 -0500	{}		{"Result":"My test output"}

![image](https://github.com/elsa-workflows/elsa-core/assets/167574820/72c228f3-78a3-4fd3-a307-040acaa378db)

7. I am able to define and run workflow through code below with no issues

        var commitWave = new CommitWaveActivity();
        var errorHandler = new ErrorHandlerActivity();
        var start = new ActivityWithOutput();

        var decide = new If(context => context.GetLastResult<string>() == "test")
        {
            Then = errorHandler,
            Else = commitWave
        };

        var workflow = new Workflow
        {
            Root = new Flowchart
            { 
                Activities =
                    {
                       start,
                        decide,
                        commitWave,
                        errorHandler
                    },
                Connections =
                {
                new Connection(start, decide) 

            },
                RunAsynchronously = true
            }
        };

        var workflowRunner = context.ServiceProvider.GetRequiredService<IWorkflowRunner>();
        // Run the workflow.
        var result = await workflowRunner.RunAsync(workflow, new RunWorkflowOptions
        {
            CancellationTokens = cancellationToken 
        }, cancellationToken).ConfigureAwait(false);

![image](https://github.com/elsa-workflows/elsa-core/assets/167574820/7527d247-7b9e-4c7a-8c37-6967941c2de3)

**Workflow **

{
	"id": "d1018d675957eb79",
	"definitionId": "f98f55f53525131f",
	"name": "Workflow 1",
	"createdAt": "2024-04-26T18:34:21.542086+00:00",
	"version": 1,
	"toolVersion": "3.1.0.0",
	"variables": [
		{
			"id": "51c60a0a2c474dc3",
			"name": "TestVariable",
			"typeName": "String",
			"isArray": false,
			"value": "",
			"storageDriverTypeName": "Elsa.Workflows.Services.MemoryStorageDriver, Elsa.Workflows.Core"
		}
	],
	"inputs": [],
	"outputs": [],
	"outcomes": [],
	"customProperties": {
		"logPersistenceMode": {
			"default": "Include"
		}
	},
	"isReadonly": false,
	"isSystem": false,
	"isLatest": true,
	"isPublished": true,
	"options": {
		"autoUpdateConsumingWorkflows": false
	},
	"root": {
		"type": "Elsa.Flowchart",
		"version": 1,
		"id": "fd83b3311f52de3a",
		"nodeId": "Workflow1:fd83b3311f52de3a",
		"metadata": {},
		"customProperties": {
			"source": "FlowchartJsonConverter.cs:45",
			"notFoundConnections": [],
			"canStartWorkflow": false,
			"runAsynchronously": false
		},
		"activities": [
			{
				"result": {
					"typeName": "String",
					"memoryReference": {
						"id": "51c60a0a2c474dc3"
					}
				},
				"id": "a039262aee005352",
				"nodeId": "Workflow1:fd83b3311f52de3a:a039262aee005352",
				"name": "ActivityWithOutput1",
				"type": "MyCustomActivity.ActivityWithOutput",
				"version": 1,
				"customProperties": {
					"canStartWorkflow": false,
					"runAsynchronously": false
				},
				"metadata": {
					"designer": {
						"position": {
							"x": -622.5,
							"y": 284
						},
						"size": {
							"width": 174.46875,
							"height": 50
						}
					}
				}
			},
			{
				"condition": {
					"typeName": "Boolean",
					"expression": {
						"type": "CSharp",
						"value": "((string)Variables.Get(\"TestVariable\")) == \"test\""
					}
				},
				"then": {
					"type": "Elsa.Flowchart",
					"version": 1,
					"id": "c03bdd074c22acbc",
					"nodeId": "Workflow1:fd83b3311f52de3a:d66b4afa41299c84:c03bdd074c22acbc",
					"metadata": {},
					"customProperties": {
						"source": "FlowchartJsonConverter.cs:45",
						"notFoundConnections": [],
						"canStartWorkflow": false,
						"runAsynchronously": false
					},
					"activities": [
						{
							"result": null,
							"id": "8da4cf92fa1f60",
							"nodeId": "Workflow1:fd83b3311f52de3a:d66b4afa41299c84:c03bdd074c22acbc:8da4cf92fa1f60",
							"name": "CommitWaveActivity1",
							"type": "MyCustomActivity.CommitWaveActivity1",
							"version": 1,
							"customProperties": {
								"canStartWorkflow": false,
								"runAsynchronously": false
							},
							"metadata": {
								"designer": {
									"position": {
										"x": -583.5,
										"y": -145
									},
									"size": {
										"width": 189.625,
										"height": 50
									}
								}
							}
						}
					],
					"connections": []
				},
				"else": {
					"type": "Elsa.Flowchart",
					"version": 1,
					"id": "ffbe2b28f6b85397",
					"nodeId": "Workflow1:fd83b3311f52de3a:d66b4afa41299c84:ffbe2b28f6b85397",
					"metadata": {},
					"customProperties": {
						"source": "FlowchartJsonConverter.cs:45",
						"notFoundConnections": [],
						"canStartWorkflow": false,
						"runAsynchronously": false
					},
					"activities": [
						{
							"result": null,
							"id": "bbe7a7414fb4ce31",
							"nodeId": "Workflow1:fd83b3311f52de3a:d66b4afa41299c84:ffbe2b28f6b85397:bbe7a7414fb4ce31",
							"name": "ErrorHandlerActivity1",
							"type": "MyCustomActivity.ErrorHandlerActivity",
							"version": 1,
							"customProperties": {
								"canStartWorkflow": false,
								"runAsynchronously": false
							},
							"metadata": {
								"designer": {
									"position": {
										"x": -514.5,
										"y": -78
									},
									"size": {
										"width": 184.625,
										"height": 50
									}
								}
							}
						}
					],
					"connections": []
				},
				"result": null,
				"id": "d66b4afa41299c84",
				"nodeId": "Workflow1:fd83b3311f52de3a:d66b4afa41299c84",
				"name": "If1",
				"type": "Elsa.If",
				"version": 1,
				"customProperties": {
					"canStartWorkflow": false,
					"runAsynchronously": false
				},
				"metadata": {
					"designer": {
						"position": {
							"x": -140,
							"y": 300
						},
						"size": {
							"width": 376.453125,
							"height": 120
						}
					}
				}
			}
		],
		"connections": [
			{
				"source": {
					"activity": "a039262aee005352",
					"port": "Done"
				},
				"target": {
					"activity": "d66b4afa41299c84",
					"port": "In"
				}
			}
		]
	}
}

**CustomActivity - ActivityWithOutput1**

    public class ActivityWithOutput : CodeActivity<string>
    { 

        protected override void Execute(ActivityExecutionContext context)
        { 
            var result = "My test output";
            context.SetResult(result);

        }
    }

**Comment by glime-ai** [comment]
Based on the context provided, it seems like you are encountering an issue with accessing the output of a custom activity (`ActivityWithOutput1`) in an Elsa workflow. The error message you're receiving indicates that there is a problem with the memory block reference for a variable or output.

Here are a few things to check and try:

1. **Variable Initialization**: Ensure that the variable `TestVariable` is properly initialized and accessible within the workflow context. It should be defined in the workflow variables and should be set with the output of `ActivityWithOutput1`.

2. **Workflow JSON**: Verify that the workflow JSON is correctly structured and that the variable `TestVariable` is correctly referenced in the `If` activity's condition. The JSON should accurately represent the workflow structure, including variables and their connections.

3. **Custom Activity Output**: Make sure that `ActivityWithOutput1` is correctly setting its output using `context.SetResult(result);` and that the result is of the expected type.

4. **Workflow Execution**: When executing the workflow from the JSON definition, ensure that the workflow runner is properly configured and that the necessary services are registered, especially if custom activities are involved.

5. **Elsa Version**: Check if you are using a compatible version of Elsa with your .NET 8 application. There might be breaking changes or updates that affect how workflows are executed or how variables are handled.

6. **Debugging**: Use debugging tools to step through the workflow execution and inspect the workflow context at runtime. This can help identify where the variable or output is not being set or accessed correctly.

7. **Elsa Documentation and Community**: Refer to the official Elsa documentation and community forums for guidance on defining and using variables, custom activities, and handling outputs. There might be known issues or updates that address your problem.

8. **Elsa GitHub Issues**: Since you are using custom activities and encountering a specific error, it might be helpful to check the Elsa GitHub issues page for similar problems reported by other users or to open a new issue with detailed information about your case.

If after checking these points the issue persists, please provide additional details such as the full error message, the version of Elsa you are using, and any relevant code snippets or configuration details.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Reply by mremadevi** [reply]
#1 - As said in detail , the variable itself is not showing up when loading from  JSON
